File: Java/CANifier Demo/src/org/usfirst/frc/team3539/robot/Framework/ConcurrentScheduler.java
package org.usfirst.frc.team3539.robot.Framework;

import com.ctre.phoenix.ILoopable;
import java.util.ArrayList;

public class ConcurrentScheduler implements com.ctre.phoenix.ILoopable {
	ArrayList<com.ctre.phoenix.ILoopable> _loops = new ArrayList<com.ctre.phoenix.ILoopable>();
	ArrayList<Boolean> _enabs = new ArrayList<Boolean>();

	public ConcurrentScheduler() {
	}

	public void add(ILoopable newLoop, boolean bEnabled) {
		for (ILoopable loop : _loops) {
			if (loop == newLoop)
				return; /* Loop exists, ignore it */
		}
		_loops.add(newLoop);
		_enabs.add(bEnabled);

		if (bEnabled)
			start(newLoop);
	}

	/* Method overloaded to provide a default state */
	public void add(ILoopable newLoop) {
		for (ILoopable loop : _loops) {
			if (loop == newLoop)
				return; /* Loop exists, ignore it */
		}
		_loops.add(newLoop);
		_enabs.add(true);

		if (true)
			start(newLoop);
	}

	public void start(ILoopable toStart) {
		for (int i = 0; i < _loops.size(); ++i) {
			ILoopable lp = (ILoopable) _loops.get(i);

			if (lp == toStart) {
				lp.onStart();
				_enabs.set(i, true);

				return;
			}
		}
		System.out.format("CTR: Could not find object in scheduler");
	}

	public void stop(ILoopable toStart) {
		for (int i = 0; i < _loops.size(); ++i) {
			ILoopable lp = (ILoopable) _loops.get(i);
			boolean en = (boolean) _enabs.get(i);

			if (lp == toStart) {
				if (en == true) {
					lp.onStop();
					_enabs.set(i, false);
				}
				return;
			}
		}
		System.out.format("CTR: Could not find object in scheduler");
	}

	public void removeAll() {
		_loops.clear();
		_enabs.clear();
	}

	public void startAll() {
		for (int i = 0; i < _loops.size(); ++i) {
			ILoopable lp = (ILoopable) _loops.get(i);

			lp.onStart();
			_enabs.set(i, true);
		}
	}

	public void stopAll() {
		for (int i = 0; i < _loops.size(); ++i) {
			ILoopable lp = (ILoopable) _loops.get(i);

			lp.onStop();
			_enabs.set(i, false);
		}
	}

	public void process() {
		for (int i = 0; i < _loops.size(); ++i) {
			ILoopable lp = (ILoopable) _loops.get(i);
			boolean en = (boolean) _enabs.get(i);

			if (en) {
				lp.onLoop();
			} else {
				/*
				 * Current ILoopable is turned off, don't call OnLoop() for it
				 */
			}
		}
	}

	// --- ILoopable ---/
	public void onStart() {
		startAll();
	}

	public void onLoop() {
		process();
	}

	public boolean isDone() {
		return false;
	}

	public void onStop() {
		stopAll();
	}
}

File: HERO C#/RC Mecanum Bot/Framework/ConcurrentScheduler.cs
using System;
using Microsoft.SPOT;

namespace CTRE.Phoenix.Tasking
{
    public class ConcurrentScheduler : IProcessable, ILoopable
    {
        System.Collections.ArrayList _loops = new System.Collections.ArrayList();
        System.Collections.ArrayList _enabs = new System.Collections.ArrayList();

        int _periodMs;
        PeriodicTimeout _timeout;

        public ConcurrentScheduler(int periodMs)
        {
            _periodMs = periodMs;
            _timeout = new PeriodicTimeout(periodMs);
        }
        public void Add(ILoopable newLoop)
        {
            foreach (var loop in _loops)
            {
                if (loop == newLoop)
                    return; /* already here */
            }
            _loops.Add(newLoop);
            _enabs.Add(true);
        }

        public void Start(ILoopable toStart)
        {
            for (int i = 0; i < _loops.Count; ++i)
            {
                ILoopable lp = (ILoopable)_loops[i];
                bool en = (bool)_enabs[i];

                if (lp == toStart)
                {
                    if (en == false)
                    {
                        _enabs[i] = true;
                        lp.OnStart();
                    }
                    return;
                }
            }
            Debug.Print("CTR: Could not find object in scheduler");
        }
        public void Stop(ILoopable toStart)
        {
            for (int i = 0; i < _loops.Count; ++i)
            {
                ILoopable lp = (ILoopable)_loops[i];
                bool en = (bool)_enabs[i];

                if (lp == toStart)
                {
                    if (en == true)
                    {
                        _enabs[i] = false;
                        lp.OnStop();
                    }
                    return;
                }
            }
            Debug.Print("CTR: Could not find object in scheduler");
        }

        public void RemoveAll()
        {
            _loops.Clear();
            _enabs.Clear();
        }

        public void StartAll()
        {
            foreach (ILoopable lp in _loops)
            {
                lp.OnStart();
            }
        }
        public void StopAll()
        {
            foreach (ILoopable lp in _loops)
            {
                lp.OnStop();
            }
        }
        public void Process()
        {
            if (_timeout.Process())
            {
                for (int i = 0; i < _loops.Count; ++i)
                {
                    ILoopable lp = (ILoopable)_loops[i];
                    bool en = (bool)_enabs[i];

                    if (en)
                    {
                        lp.OnLoop();
                    }
                    else
                    {
                        /* this loopable is turned off, don't call it */
                    }
                }
            }
        }
        //--- Loopable ---/
        public void OnStart()
        {
            StartAll();
        }

        public void OnLoop()
        {
            Process();
        }

        public bool IsDone()
        {
            /* TODO poll all the tasks, if all tasks are enabled and done, then return true */
            return false;
        }

        public void OnStop()
        {
            StopAll();
        }
    }
}